### ISSUE TITLE
Backend tests failing

### ISSUE THREAD

**Issue by raybellis** [ether/etherpad-lite#3567]
Something has changed in my environment such that I'm getting 7 back end test failures even on the release 1.7.0 code (I originally thought I'd broken something in my async branch).

The first reported error is:
```
  setText
    1) Sets text on a pad Id
superagent: double callback bug
```
with log messages:
```
[2019-02-28 17:52:30.905] [INFO] console -   1) setText
       Sets text on a pad Id:
     TypeError: Cannot read property 'body' of undefined
      at /home/ray/etherpad-old/tests/backend/specs/api/pad.js:392:14
      at Test._assertFunction (node_modules/supertest/lib/test.js:281:11)
      at Test.assert (node_modules/supertest/lib/test.js:171:18)
      at assert (node_modules/supertest/lib/test.js:131:12)
      at /home/ray/etherpad-old/src/node_modules/supertest/lib/test.js:128:5
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:728:3)
      at ClientRequest.req.once.err (node_modules/superagent/lib/node/index.js:647:10)
      at Socket.socketErrorListener (_http_client.js:392:9)
      at emitErrorNT (internal/streams/destroy.js:82:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
      at process._tickCallback (internal/process/next_tick.js:63:19)
```

I'm suspecting that an upstream module has changed in an incompatible way.

@muxator Are you able to replicate this?

**Comment by muxator** [comment]
Hi Ray,

node **8.9.0**, linux, default `settings.json`.

- **1.7.0**: backend tests pass (details below)
- **1.7.5**: backend tests pass (details below)
- **async-PR**: three failures (the same happens on `async`, details below)

At first I have run everything "by hand". After seeing that our results were different, I've redone everything via a script. Hopefully you could replicate the results.

Execution logs for **1.7.0**:
```
[...]
[2019-03-01 02:10:42.179] [INFO] console -   75 passing (927ms)
[2019-03-01 02:10:42.179] [INFO] console -   2 pending
[...]
```

Execution logs for **1.7.5**:
```
[2019-03-01 02:12:26.744] [INFO] console -   75 passing (931ms)
[2019-03-01 02:12:26.744] [INFO] console -   2 pending
```

Execution logs for current `async-PR` (950eae55d440), same results for `async` (64fdc73cdc47):
```
[2019-03-01 02:13:56.359] [INFO] console -   1) createSession
       Creates a session for an Author:
     Error: Unable to create Session
      at <BASEDIR>/tests/backend/specs/api/sessionsAndGroups.js:174:65             
      at Test._assertFunction (node_modules/supertest/lib/test.js:281:11)                                                     
      at Test.assert (node_modules/supertest/lib/test.js:171:18)                                                              
      at assert (node_modules/supertest/lib/test.js:131:12)                                                                   
      at <BASEDIR>/src/node_modules/supertest/lib/test.js:128:5                    
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:716:12)                                             
      at parser (node_modules/superagent/lib/node/index.js:916:18)                                                            
      at IncomingMessage.res.on (node_modules/superagent/lib/node/parsers/json.js:19:7)                                       
      at endReadableNT (_stream_readable.js:1056:12)                                                                          
      at _combinedTickCallback (internal/process/next_tick.js:138:11)                                                         
      at process._tickCallback (internal/process/next_tick.js:180:9)                                                          
                                                                                                                              
[2019-03-01 02:13:56.360] [INFO] console -   2) getSessionInfo
       Gets session inf:
     Error: Unable to get Session info
      at <BASEDIR>/tests/backend/specs/api/sessionsAndGroups.js:186:119            
      at Test._assertFunction (node_modules/supertest/lib/test.js:281:11)                                                     
      at Test.assert (node_modules/supertest/lib/test.js:171:18)                                                              
      at assert (node_modules/supertest/lib/test.js:131:12)                                                                   
      at <BASEDIR>/src/node_modules/supertest/lib/test.js:128:5                    
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:716:12)                                             
      at parser (node_modules/superagent/lib/node/index.js:916:18)                                                            
      at IncomingMessage.res.on (node_modules/superagent/lib/node/parsers/json.js:19:7)                                       
      at endReadableNT (_stream_readable.js:1056:12)                                                                          
      at _combinedTickCallback (internal/process/next_tick.js:138:11)                                                         
      at process._tickCallback (internal/process/next_tick.js:180:9)                                                          
                                                                                                                              
[2019-03-01 02:13:56.360] [INFO] console -   3) deleteSession
       Deletes a session:
     Error: Unable to delete a session
      at <BASEDIR>/tests/backend/specs/api/sessionsAndGroups.js:208:37             
      at Test._assertFunction (node_modules/supertest/lib/test.js:281:11)                                                     
      at Test.assert (node_modules/supertest/lib/test.js:171:18)                                                              
      at assert (node_modules/supertest/lib/test.js:131:12)                                                                   
      at <BASEDIR>/src/node_modules/supertest/lib/test.js:128:5                    
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:716:12)                                             
      at parser (node_modules/superagent/lib/node/index.js:916:18)                                                            
      at IncomingMessage.res.on (node_modules/superagent/lib/node/parsers/json.js:19:7)                                       
      at endReadableNT (_stream_readable.js:1056:12)                                                                          
      at _combinedTickCallback (internal/process/next_tick.js:138:11)                                                         
      at process._tickCallback (internal/process/next_tick.js:180:9)                                                          
```


## How to replicate

```bash
$ hg up -r <VERSION>    # version 1.7.0|1.7.5|async-PR. The git equivalent should be "git checkout <VERSION>"
$ ./clean.sh            # cleanup the environment, see later
$ bin/installDeps.sh    # download dependencies

# one-liner made up for easing your replication (please tolerate the ugly "sleep" statement)
$ node node_modules/ep_etherpad-lite/node/server.js & (sleep 7 ; cd src ; npm test) ; fg
```

Contents of `clean.sh` (put it in `<BASEDIR>`):
```bash
#!/bin/bash
#
# Cleans up Etherpad installation

set -eu

DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null && pwd )"
ETHERPAD_PORT=9001

rm -rf "${DIR}/src/node_modules"
rm -rf "${DIR}/node_modules"
rm -f "${DIR}/package-lock.json"
rm -f "${DIR}/src/package-lock.json"
rm -f "${DIR}/var/dirty.db"
rm -f "${DIR}/src/.ep_initialized"
```

**Comment by raybellis** [comment]
Ah-ha!  If I run EP 7.x with Node 8.9.0 it works.   If I run with NodeJS 8.15.0 it doesn't.  I'll attempt to bisect on the node version to confirm what that is.

In the meantime, your test failures with the async branches are resolved in commit d03da2c8eb36f94a53621566e7cc0cfef8a493e5.

I'm still investigating the ones that

**Comment by raybellis** [comment]
I've bisected the regression as starting with NodeJS 8.14.0.   It's OK with NodeJS 8.13.0 (which was the only 8.13 release).

**Comment by Wikinaut** [comment]
* 78c057af3117045e517b32f87e86bacd61a9e937 works with node 11.10.1
* d03da2c8eb36f94a53621566e7cc0cfef8a493e5 (async) works with node 11.10.1

**Comment by Wikinaut** [comment]
@muxator May I suggest that you commit your ```clean.sh``` (as shown above) script to the ```/bin``` directory in the repo as a maintenance script? What do you/the others think?

**Comment by raybellis** [comment]
@Wikinaut it passes all 75 backend tests for those versions for you?   I just tried 11.10.1 here and it still failed.   OTOH, I did a git bisect and got this, from between 8.13.0 and 8.14.0

```
93dba83fb0fb46ee2ea87163f435392490b4d59b is the first bad commit
commit 93dba83fb0fb46ee2ea87163f435392490b4d59b
Author: Matteo Collina <hello@matteocollina.com>
Date:   Tue Aug 21 17:26:51 2018 +0200

    deps,http: http_parser set max header size to 8KB
    
    CVE-2018-12121
 ```

**Comment by Wikinaut** [comment]
@raybellis What do I have to do exactly?

**Comment by raybellis** [comment]
You said "works with 11.1.0", but I think you're conflating testing the async work (per the other ticket) with this one which was intended to be specifically about back end tests that are failing even on the `develop` branch (those tests done by going into the `src/` folder and running `npm test`, with the server running, as described above by @muxator)

**Comment by raybellis** [comment]
The referenced Github commit (
93dba83fb0fb46ee2ea87163f435392490b4d59b) prevents `GET` data from being submitted with URL parameters that exceed 8kB.   The back end tests appear to exceed that.

**Comment by raybellis** [comment]
The implication of this is that some methods in the HTTP API will become unusable for larger pads unless they migrate to `POST` for bulk content instead of `GET` (whilst perhaps still using URL parameters for the non-bulk parameters such as the API key)

**Comment by muxator** [comment]
A recap to keep track of what's going on.

1. Three failures in backend tests on sessions. They were due to to stricter checks introduced in https://github.com/ether/etherpad-lite/commit/37e4081307041f7c7edf886a46a37f780960fb4f#diff-c6aae0f724e78a93dd699ad297010f52R89. When using dirtydb, the DB layer may return `undefined`, where mysql would return `null`. This seems a ueberDB2 bug. Nonetheless Ray fixed restoring the older, lax behaviour in d03da2c8eb36. This modification was included in async-PR (46c8a7bbbc4d).
2. incompatibility with nodejs >= 8.14. Ray opened a dedicated issue, #3568.

**Comment by muxator** [comment]
I guess this can be closed (see previous comment for explanation).

There is a remaining issue with the HTTP API, and for that we have a dedicated ticket (#3568).
